From patchwork Tue Mar 19 22:19:56 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: David Rientjes <rientjes@google.com>
X-Patchwork-Id: 10860535
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 480501823
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue, 19 Mar 2019 22:20:01 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 2A3DC296BD
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue, 19 Mar 2019 22:20:01 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 1A692296BF; Tue, 19 Mar 2019 22:20:01 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-15.5 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI,
	USER_IN_DEF_DKIM_WL autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id B7BF5296BD
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue, 19 Mar 2019 22:20:00 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1727434AbfCSWT7 (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Tue, 19 Mar 2019 18:19:59 -0400
Received: from mail-pf1-f193.google.com ([209.85.210.193]:37768 "EHLO
        mail-pf1-f193.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1727418AbfCSWT6 (ORCPT <rfc822;kvm@vger.kernel.org>);
        Tue, 19 Mar 2019 18:19:58 -0400
Received: by mail-pf1-f193.google.com with SMTP id 8so377199pfr.4
        for <kvm@vger.kernel.org>; Tue, 19 Mar 2019 15:19:57 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=date:from:to:cc:subject:message-id:user-agent:mime-version;
        bh=59RGI7G7pmr5cNfnWSfGjhKSPq5Y0VsvQ6GYsDbuqfs=;
        b=XfpYf4gL1c44dexWEA7/hn7uwvJMen3gV457ORsj+K56A2l9K4zVq4d0xAt/eV5fYQ
         FFdfSejAoM/wUrBF0katomC+jxmtew9xvorimKgXDejrEiSiepz/JtHerg1LrEPRXy4E
         7VBIREzygXHHEPHy6zske8Spvrk74fwKoySsY0vNLW4CVvJSH/KXkSwfdPD7kinZQwvU
         md4DMtxPynIkLB5XqaaMmNfWNvtzUy0u2tzr+8YYEPVljgDTX48TmK8eRv3RPpbCghVp
         cdCq+FgXDMTifYwo6k5tpotrI0haDb9mUKhNUr02JfD9Hjb9RYMT47nb8X5PJKbqnB+4
         za4w==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:date:from:to:cc:subject:message-id:user-agent
         :mime-version;
        bh=59RGI7G7pmr5cNfnWSfGjhKSPq5Y0VsvQ6GYsDbuqfs=;
        b=KKEv7zBdfSXfQpHBU1LdJDgjkT7frvHE0aSFRESwZdyRIlFcPACJQN+N0E1oAbA1v+
         XvquRyC/noZGGHyXjSqDp0gOgv/6rewsZ/NtdOcbmlkX9TJmchKURA5+P46fq0xJm8vw
         Y1wBrIcFDLKwcF+GIup2SOM6SUSZTsbqJq9IEtG6k1odEabQ4ElKpf01k6YnPHmT2BKW
         uUt9WhMfOg6cCJ3pnZ6XRfttjx5yHwhmItSSRzjvhytakas1rOPzmkuXsSLO9PHC3Fur
         VvFJX+hXJ/n6NDUPbQkMOgI76ykFLX1mSNN3+8SzmBfX8WY3MscQEy8EDmZxNtmIErtw
         gEYA==
X-Gm-Message-State: APjAAAU0TnlebgRXDgFDBKMWgnWklrZmaoqlAmGqoEznS/it2eN8n/9F
        8dSNrQroCx6WYLqrPM8g62P7gQ==
X-Google-Smtp-Source: 
 APXvYqxU8lzqwVXJMTs59fVnzSRbf+8jxap77aX/EoV/+K6jh4ghRHHZhbPshUAGv7LFYZRYATv3gw==
X-Received: by 2002:a17:902:14b:: with SMTP id
 69mr4383333plb.216.1553033997376;
        Tue, 19 Mar 2019 15:19:57 -0700 (PDT)
Received: from [2620:15c:17:3:3a5:23a7:5e32:4598]
 ([2620:15c:17:3:3a5:23a7:5e32:4598])
        by smtp.gmail.com with ESMTPSA id j1sm15846pgp.84.2019.03.19.15.19.56
        (version=TLS1_2 cipher=ECDHE-RSA-CHACHA20-POLY1305 bits=256/256);
        Tue, 19 Mar 2019 15:19:56 -0700 (PDT)
Date: Tue, 19 Mar 2019 15:19:56 -0700 (PDT)
From: David Rientjes <rientjes@google.com>
X-X-Sender: rientjes@chino.kir.corp.google.com
To: Paolo Bonzini <pbonzini@redhat.com>,
 =?utf-8?b?UmFkaW0gS3LEjW3DocWZ?= <rkrcmar@redhat.com>,
 Joerg Roedel <joro@8bytes.org>
cc: Cfir Cohen <cfir@google.com>,
        Brijesh Singh <brijesh.singh@amd.com>,
        Tom Lendacky <thomas.lendacky@amd.com>, x86@kernel.org,
        kvm@vger.kernel.org, linux-kernel@vger.kernel.org
Subject: [patch] kvm: svm: fix potential get_num_contig_pages overflow
Message-ID: <alpine.DEB.2.21.1903191517380.256304@chino.kir.corp.google.com>
User-Agent: Alpine 2.21 (DEB 202 2017-01-01)
MIME-Version: 1.0
Content-Type: text/plain; charset=US-ASCII
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

get_num_contig_pages() could potentially overflow int so make its type 
consistent with its usage.

Reported-by: Cfir Cohen <cfir@google.com>
Signed-off-by: David Rientjes <rientjes@google.com>
---
 arch/x86/kvm/svm.c | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

diff --git a/arch/x86/kvm/svm.c b/arch/x86/kvm/svm.c
--- a/arch/x86/kvm/svm.c
+++ b/arch/x86/kvm/svm.c
@@ -6422,11 +6422,11 @@ static int sev_launch_start(struct kvm *kvm, struct kvm_sev_cmd *argp)
 	return ret;
 }
 
-static int get_num_contig_pages(int idx, struct page **inpages,
-				unsigned long npages)
+static unsigned long get_num_contig_pages(unsigned long idx,
+				struct page **inpages, unsigned long npages)
 {
 	unsigned long paddr, next_paddr;
-	int i = idx + 1, pages = 1;
+	unsigned long i = idx + 1, pages = 1;
 
 	/* find the number of contiguous pages starting from idx */
 	paddr = __sme_page_pa(inpages[idx]);
@@ -6445,12 +6445,12 @@ static int get_num_contig_pages(int idx, struct page **inpages,
 
 static int sev_launch_update_data(struct kvm *kvm, struct kvm_sev_cmd *argp)
 {
-	unsigned long vaddr, vaddr_end, next_vaddr, npages, size;
+	unsigned long vaddr, vaddr_end, next_vaddr, npages, pages, size, i;
 	struct kvm_sev_info *sev = &to_kvm_svm(kvm)->sev_info;
 	struct kvm_sev_launch_update_data params;
 	struct sev_data_launch_update_data *data;
 	struct page **inpages;
-	int i, ret, pages;
+	int ret;
 
 	if (!sev_guest(kvm))
 		return -ENOTTY;
